rfcs/004-typechecking.md

```diff
@@ -0,0 +1,496 @@
+
+As long as there is no `Dyn` involved by the type given by either the signature
+of a function or the annotation of an argument, we want subtyping to not change
+the current behavior. For the typechecker, this means following this guideline:
+
+_Never infer a `Dyn` type implicitly. `Dyn`, and dictionary subtyping, must
+always stem from an explicit annotation._
+
+For example, the following program is expected to **fail** typechecking:
+
+```nickel
+let uni_pair : forall a. a -> a -> {fst: a, snd: a} = fun x y =>
+  { fst = x, snd = y } in
+
+uni_pair 1 "a" : _
+```
+
+We could infer the type `Dyn` for the instantiation of `a`. But this is
+currently an error, and it should arguably stay this way. `Dyn` can make a lot
+of such cases pass, only to manifest as a later type error – and probably a
+disconcerting one involving a `Dyn` coming from nowhere – when the result of the
+function is actually used. However, if one of the argument is explicitly of type
+`Dyn`, then it is reasonable to allow the other one to be silently upcast:
+
+```nickel
+let foo : Dyn = 1 in
+uni_pair foo 1 : _
+```
+
+Unfortunately, we don't have yet a good characterization of this as a property
+of the declarative type system or type inference . It is an informal guideline,
+which is only reflected in the algorithmic type inference for now.
+
+### Polymorphism and subtyping
+
+The previous example shows an interesting aspect of combining subtyping and
+polymorphism: if we instantiate type variables at the first type we see, as we
+do currently, typechecking is sensitive to the order of arguments:
+
+```nickel
+# ok, a is instantiated to Dyn, and then 1 : Dyn
+uni_pair foo 1 : _
+# would fail, a is instantiated to Num, but foo is not of type Num
+uni_pair 1 foo : _
+```
+
+This problem is dubbed the Tabby-first problem in [the review of bidirectional
+typing of Dunfield and Krishnaswami](https://arxiv.org/abs/1908.05839).
+Surprisingly, it doesn't seem to have been deeply studied (outside of the
+subtyping relation induced by polymorphism). For subtyping induced by
+polymorphism, there are different trade-offs: Dunfield and Krishnaswami
+restricts their system to impredicative polymorphism, which is then
+order-independent [^3]. In exchange, the arrow type has the expected variance
+(that is, eta-expansion preserves typeability). On the other hand, Haskell
+deemed impredicative polymorphism to be more important and chose to abandon the
+variance of arrows and other type constructors. Coupled with a so-called
+QuickLook phase, Haskell is able to most common impredicative instantiations.
+
+We can't apply the Haskell approach naively, because in the presence of `Dyn`,
+an impredicative instantiation is not the only possibility anymore, even if we
+don't do deep instantiation or deep skolemization:
+
+```nickel
+let foo : Dyn = null in
+let id : forall a. a -> a = fun x => x in
+let concat : forall b. Array b -> Array b -> Array b = (@) in
+
+concat [id] [null]
```

Comment:
I don't this example is correct. This you can absolutely instantiate `concat [id]` with `b = Bool -> Bool`.

An example that would work is, I believe
```suggestion
let foo : Dyn = null in
let aid : Array (forall a. a -> a) = [fun x => x] in
let concat : forall b. Array b -> Array b -> Array b = (@) in

concat aid [null]
```

Reply:
Ah, because the typechecker would eagerly instantiate `id` to `?a -> ?a` when writing `[id]`, giving `[?a -> ?a]`, allowing it to be unified with a concrete type?

Reply:
Right.